github.go

```diff
@@ -83,48 +90,181 @@ func (p *githubChangeProcessor) process(c *change) error {
+		return pullRequestInfo{}, err
 	}
-	req.Header.Add("Accept", "application/vnd.github.v3+json")
+	req.Header.Add("Accept", "application/vnd.github+json")
+	req.Header.Add("X-GitHub-Api-Version", "2022-11-28")
 	if user, token := os.Getenv("GITHUB_ACTOR"), os.Getenv("GITHUB_TOKEN"); user != "" && token != "" {
 		req.SetBasicAuth(user, token)
 	}
 
 	resp, err := http.DefaultClient.Do(req)
 	if err != nil {
-		return "", err
+		return pullRequestInfo{}, err
 	}
 	defer resp.Body.Close()
 
 	if resp.StatusCode >= 400 {
 		if resp.StatusCode >= 403 {
 			logrus.Warn("Forbidden response, try setting GITHUB_USER and GITHUB_TOKEN environment variables")
 		}
-		return "", fmt.Errorf("unexpected status code %d for %s", resp.StatusCode, u)
+		return pullRequestInfo{}, fmt.Errorf("unexpected status code %d for %s", resp.StatusCode, u)
 	}
 
 	dec := json.NewDecoder(resp.Body)
 
-	pr := struct {
-		Title string `json:"title"`
-	}{}
-	if err := dec.Decode(&pr); err != nil {
-		return "", err
+	var info pullRequestInfo
+	if err := dec.Decode(&info); err != nil {
+		return pullRequestInfo{}, err
+	}
+	if info.Title == "" {
+		return pullRequestInfo{}, fmt.Errorf("unexpected empty title for %s", u)
+	}
+
+	cacheB, err := json.Marshal(info)
+	if err == nil {
+		p.cache.Put(key, cacheB)
+	}
+
+	return info, nil
+}
+
+func (p *githubChangeProcessor) advisoryChange(c *change, info advisoryInfo, ghsa string) {
+	c.IsSecurity = true
+	c.Link = info.Link
+	if c.Link == "" {
+		c.Link = fmt.Sprintf("https://github.com/%s/security/advisories/%s", p.repo, ghsa)
+	}
+	summary := info.Summary
+	if summary == "" {
+		summary = "Github Security Advisory"
+	}
+	c.Formatted = fmt.Sprintf("%s [%s](%s)", summary, ghsa, c.Link)
+	cveInfo := []string{}
+	if info.CVE != "" {
+		cveInfo = append(cveInfo, info.CVE)
+	}
+	if info.Severity != "" {
+		cveInfo = append(cveInfo, info.Severity)
+	}
+	if len(cveInfo) > 0 {
+		prefix := "[" + strings.Join(cveInfo, ", ") + "] "
+		c.Formatted = prefix + c.Formatted
 	}
-	if pr.Title == "" {
-		return "", fmt.Errorf("unexpected empty title for %s", u)
+}
+
+type advisoryInfo struct {
+	CVE         string `json:"cve_id"`
+	Link        string `json:"html_url"`
+	Summary     string `json:"summary"`
+	Description string `json:"description"`
+	Severity    string `json:"severity"`
+}
+
+// getAdvisoryInfo returns github security advisory info
+//
+// See https://docs.github.com/en/rest/security-advisories/repository-advisories?apiVersion=2022-11-28#get-a-repository-security-advisory
+func (p *githubChangeProcessor) getAdvisoryInfo(repo, advisory string) (advisoryInfo, error) {
+	u := fmt.Sprintf("https://api.github.com/repos/%s/security-advisories/%s", repo, advisory)
+	key := u + " cve link summary description severity"
+	if !p.refreshCache {
+		if b, ok := p.cache.Get(key); ok {
+			var info advisoryInfo
+			if err := json.Unmarshal(b, &info); err == nil {
+				return info, nil
+			}
+		}
+	}
+	req, err := http.NewRequest("GET", u, nil)
```

Comment:
minor and probably a follow-on cleanup, but this is the 2nd instance of setting up an HTTP API call to GH with API version, auth, error handling; might be worth extracting to a function so if the headers/API version needs an update in the future it's easy to do in one place.

Reply:
Yeah we can, originally they were slightly different because the API docs seemed to be updated since I started working on it. Good reason to have it once.